indimail-mta-2.7/new-inject.1
.TH new-inject 1
.SH NAME
new-inject \- preprocess a mail message
.SH SYNOPSIS
.B new-inject
[
.B \-nNaAhHFIMRS
]
[
.B \-f\fIsender
]
[
.I recip ...
]
.SH DESCRIPTION
With the (default)
.B \-N
option,
.B new-inject
reads a mail message from its standard input,
rewrites the message header,
and feeds the rewritten message to
.BR qmail-queue .

With the
.B \-n
option,
.B new-inject
prints the message rather than feeding it to
.BR qmail-queue .

See
.B http://pobox.com/~djb/proto/immhf.html
for a complete description of the Internet mail message header format.

The
.B FIMRS
options can be set as letters inside the
.B $QMAILINJECT
environment variable.
.SH "ADDRESS REWRITING"
.B new-inject
transforms each address according to the control files
.IR defaultdomain ,
.IR defaulthost ,
.IR me ,
.IR plusdomain ,
and
.IR rewrite .
See
.BR rewriting (5)
for further details.
If you set the environment variable
.BR $QMAILREWRITEFILE ,
and that file exists,
it overrides
.IR rewrite .
.SH "ENVELOPE ADDRESSES"
.B new-inject
builds an envelope for the message:
a list of recipient addresses that will receive the message,
and a sender address that will receive bounces.

If
.B new-inject
is printing the message rather than feeding it to
.BR qmail-queue ,
it prints an
.B Envelope-Sender
field and an 
.B Envelope-Recipients
field
at the top of the new message header.

With the
.B \-a
option,
.B new-inject
builds the envelope recipient list from the argument recipients,
i.e., all
.I recip
addresses listed on the command line.
With the
.B \-h
option,
.B new-inject
builds the envelope recipient list from the header recipients,
i.e., all addresses listed in the incoming message under
.BR Envelope-Recipients ,
or under
.BR To ,
.BR Cc ,
.BR Bcc ,
and
.B Apparently-To
if
.B Envelope-Recipients
is not supplied.
With the
.B \-H
option,
.B new-inject
uses both argument recipients and header recipients.
With the (default)
.B \-A
option,
.B new-inject
uses argument recipients,
or header recipients if there are no argument recipients.

The envelope sender is set by
the
.B \-f
option if it is supplied;
otherwise, the incoming
.B Envelope-Sender
field, if there is one;
otherwise, the incoming
.B Return-Path
field, if there is one,
and if the
.B \-S
option is not set;
otherwise, the default envelope sender, as described below.
Incoming
.B Envelope-Sender
and 
.B Return-Path
fields
are removed in any case.

The default envelope sender contains a user name,
a per-message VERP,
a per-recipient VERP,
and a host name.
The user name is set by an environment variable:
.BR $QMAILSUSER ,
.BR $QMAILUSER ,
.BR $MAILUSER ,
.BR $USER ,
or
.BR $LOGNAME ,
whichever comes first.
If the
.B \-M
option is set, the per-message VERP
contains a dash, the current time, and the process ID;
otherwise it is empty.
If the
.B \-R
option is set, the per-recipient VERP
contains a dash, the envelope recipient user name,
an equals sign, and the envelope recipient host name;
otherwise it is empty.
(If there are several recipients,
each recipient sees a different sender address.)
The host name is set by
.BR $QMAILSHOST ,
.BR $QMAILHOST ,
.BR $MAILHOST ,
or the empty string,
subject to rewriting.
.SH "RECIPIENTS"
.B new-inject
rewrites addresses in
.BR To ,
.BR Cc ,
and
.B Notice-Requested-Upon-Delivery-To
fields.
It
cleans up the spacing,
adds missing commas,
trims extra commas,
adds missing quotes,
wraps lines at a reasonable length,
converts source routes into comments,
etc.
If there are several
.B To
fields,
.B new-inject
merges them into a single
.B To
field.

.B new-inject
removes all incoming
.B Bcc
and
.B Apparently-To
fields.
.SH "SENDERS"
.B new-inject
rewrites addresses in
.B From
and
.B Sender
fields.

If the incoming message does not contain a
.B From
field,
.B new-inject
creates a new
.B From
field.

If the
.B \-F
option is set,
.B new-inject
discards any incoming
.B From
field,
and creates a new
.B From
field.

The new
.B From
field
contains a personal name, a user name, and a host name.
The personal name is set by
.BR $QMAILNAME ,
.BR $MAILNAME ,
or
.BR $NAME ,
whichever comes first;
if none of these variables are set, the personal name is omitted.
The user name is set by
.BR $QMAILUSER ,
.BR $MAILUSER ,
.BR $USER ,
or
.BR $LOGNAME .
The host name is set by
.BR $QMAILHOST ,
.BR $MAILHOST ,
or the empty string,
subject to rewriting.
.SH "RESPONSE TARGETS"
.B new-inject
rewrites
addresses in
.BR Reply-To ,
.BR Mail-Reply-To ,
and
.B Mail-Followup-To
fields.

If the environment variable
.B $QMAILMFTFILE
is set,
.B new-inject
reads a list of mailing list addresses,
one per line,
from that file.
If To+Cc includes one of those addresses (without regard to case),
and if the incoming message does not contain a
.B Mail-Followup-To
field,
.B new-inject
adds a
.B Mail-Followup-To
field
with all the To+Cc addresses.
.SH "DATE FIELDS"
If the incoming message has comprehensible
.B Date
fields,
.B new-inject
takes the last one and rewrites it in a standard format.
It transforms 2-digit (and 3-digit) years into 4-digit years.
It also transforms obsolete time zone names into numeric time zones.

Otherwise
.B new-inject
makes a new
.B Date
field with the current time in UTC, counting leap seconds.
.SH "THREADING"
If the incoming message does not contain a
.B Message-ID
field,
.B new-inject
creates a new
.B Message-ID
field.

If the
.B \-I
option is set,
.B new-inject
discards any incoming
.B Message-ID
field,
and creates a new
.B Message-ID
field.

.B new-inject
uses the control file
.I idhost
(overridden by he environment variable
.BR $QMAILIDHOST ;
defaulting to
.IR me )
in its new
.B Message-ID
field.
.I idhost
need not be the current host's name.
It is your responsibility to obtain authorization
from the owner of the
.I idhost
domain.
.SH "MISCELLANY"
.B new-inject
puts
.BR Received ,
.BR Delivered-To ,
.BR Errors-To ,
.BR Return-Receipt-To ,
.BR Resent-Sender ,
.BR Resent-From ,
.BR Resent-Reply-To ,
.BR Resent-To ,
.BR Resent-Cc ,
.BR Resent-Bcc ,
.BR Resent-Date ,
and
.B Resent-Message-ID
fields at the top of the header.
It does not rewrite any of these fields;
it follows the 822bis approach of treating
.B Resent-*
as trace fields.

.B new-inject
discards incoming
.B Content-Length
fields.
.SH "SEE ALSO"
leapsecs(3),
qmail-control(5),
rewriting(5),
qmail-queue(8)


qmail-1.03/new-inject.c
/*
 * $Log: new-inject.c,v $
 * Revision 1.9  2016-05-21 14:48:09+05:30  Cprogrammer
 * use auto_sysconfdir for leapsecs_init()
 *
 * Revision 1.8  2016-05-17 19:44:58+05:30  Cprogrammer
 * use auto_control, set by conf-control to set control directory
 *
 * Revision 1.7  2016-01-28 09:01:02+05:30  Cprogrammer
 * chdir to qmail_home before leapsecs_init()
 *
 * Revision 1.6  2010-06-08 21:59:42+05:30  Cprogrammer
 * use envdir_set() on queuedefault to set default queue parameters
 *
 * Revision 1.5  2009-08-13 19:05:56+05:30  Cprogrammer
 * code beautified
 *
 * Revision 1.4  2005-08-23 17:33:21+05:30  Cprogrammer
 * gcc 4 compliance
 *
 * Revision 1.3  2004-10-22 20:27:44+05:30  Cprogrammer
 * added RCS id
 *
 * Revision 1.2  2004-10-11 23:53:48+05:30  Cprogrammer
 * include sgetopt.h after unistd.h
 *
 * Revision 1.1  2004-06-16 01:19:55+05:30  Cprogrammer
 * Initial revision
 *
 */
#include <unistd.h>
#include "auto_sysconfdir.h"
#include "auto_control.h"
#include "env.h"
#include "envdir.h"
#include "pathexec.h"
#include "sgetopt.h"
#include "substdio.h"
#include "subfd.h"
#include "str.h"
#include "getln.h"
#include "mess822.h"
#include "strerr.h"
#include "exit.h"
#include "caltime.h"
#include "leapsecs.h"
#include "tai.h"
#include "stralloc.h"
#include "sconfig.h"
#include "case.h"
#include "constmap.h"
#include "qmail.h"
#include "sig.h"
#include "rewritehost.h"
#include "rwhconfig.h"
#include "variables.h"

#define FATAL "new-inject: fatal: "

void
nomem()
{
	strerr_die2x(111, FATAL, "out of memory");
}

config_str      qmailinject = CONFIG_STR;
config_str      name = CONFIG_STR;
config_str      user = CONFIG_STR;
config_str      host = CONFIG_STR;
config_str      suser = CONFIG_STR;
config_str      shost = CONFIG_STR;
config_str      fnmft = CONFIG_STR;
config_str      mft = CONFIG_STR;
struct constmap mapmft;
config_str      fnrewrite = CONFIG_STR;
config_str      rewrite = CONFIG_STR;
stralloc        idappend = { 0 };
struct tai      start;
stralloc        tmp = { 0 };
stralloc        tmp2 = { 0 };

void
rewritelist(list)
	stralloc       *list;
{
	if (!rewritehost_list(&tmp, list->s, list->len, config_data(&rewrite)))
		nomem();
	if (!stralloc_copy(list, &tmp))
		nomem();
}

int             recipstrategy = 'A';
int             flagkillfrom = 0;
int             flagkillmsgid = 0;
int             flagkillreturnpath = 0;
int             flagverpmess = 0;
int             flagverprecip = 0;
stralloc        sender = { 0 };
stralloc        recipients = { 0 };

void
recipient_add(addr)
	char           *addr;
{
	if (!rewritehost_addr(&tmp, addr, str_len(addr), config_data(&rewrite)))
		nomem();
	if (!stralloc_cat(&recipients, &tmp))
		nomem();
	if (!stralloc_0(&recipients))
		nomem();
}

void
recipient_addlist(list)
	stralloc       *list;
{
	int             i;
	int             j;

	for (j = i = 0; j < list->len; ++j)
	{
		if (!list->s[j])
		{
			if (list->s[i] == '+' && !stralloc_catb(&recipients, list->s + i + 1, j - i))
				nomem();
			i = j + 1;
		}
	}
}

void
sender_get(list)
	stralloc       *list;
{
	int             i;
	int             j;

	for (j = i = 0; j < list->len; ++j)
	{
		if (!list->s[j])
		{
			if (list->s[i] == '+' && !stralloc_copyb(&sender, list->s + i + 1, j - i))
				nomem();
			i = j + 1;
		}
	}
}

int             flagqueue = 1;
struct qmail    qq;

void
myput(buf, len)
	char           *buf;
	int             len;
{
	if (flagqueue)
		qmail_put(&qq, buf, len);
	else
		substdio_put(subfdoutsmall, buf, len);
}

void
myputs(buf)
	char           *buf;
{
	myput(buf, str_len(buf));
}

void
putlist(name, list)
	char           *name;
	stralloc       *list;
{
	if (!list->len)
		return;
	if (!mess822_quotelist(&tmp, list))
		nomem();
	if (!mess822_fold(&tmp2, &tmp, name, 78))
		nomem();
	myput(tmp2.s, tmp2.len);
}

mess822_time    date;
stralloc        enveloperecipients = { 0 };
int             flagenveloperecipients;
stralloc        envelopesender = { 0 };
int             flagenvelopesender;
stralloc        to = { 0 };
stralloc        cc = { 0 };
stralloc        bcc = { 0 };
stralloc        nrudt = { 0 };
stralloc        returnpath = { 0 };
stralloc        from = { 0 };
stralloc        headersender = { 0 };
stralloc        replyto = { 0 };
stralloc        mailreplyto = { 0 };
stralloc        followupto = { 0 };
stralloc        msgid = { 0 };
stralloc        top = { 0 };
stralloc        bottom = { 0 };

mess822_header  h = MESS822_HEADER;
mess822_action  a[] = {
	{"date", 0, 0, 0, 0, &date}
	, {"to", 0, 0, 0, &to, 0}
	, {"cc", 0, 0, 0, &cc, 0}
	, {"bcc", 0, 0, 0, &bcc, 0}
	, {"apparently-to", 0, 0, 0, &bcc, 0}
	, {"notice-requested-upon-delivery-to", 0, 0, 0, &nrudt, 0}
	, {"from", 0, 0, 0, &from, 0}
	, {"sender", 0, 0, 0, &headersender, 0}
	, {"reply-to", 0, 0, 0, &replyto, 0}
	, {"mail-reply-to", 0, 0, 0, &mailreplyto, 0}
	, {"mail-followup-to", 0, 0, 0, &followupto, 0}
	, {"return-path", 0, 0, 0, &returnpath, 0}
	, {"envelope-recipients", &flagenveloperecipients, 0, 0, &enveloperecipients, 0}
	, {"envelope-sender", &flagenvelopesender, 0, 0, &envelopesender, 0}
	, {"message-id", 0, &msgid, 0, 0, 0}
	, {"received", 0, &top, 0, 0, 0}
	, {"delivered-to", 0, &top, 0, 0, 0}
	, {"errors-to", 0, &top, 0, 0, 0}
	, {"return-receipt-to", 0, &top, 0, 0, 0}
	, {"resent-sender", 0, &top, 0, 0, 0}
	, {"resent-from", 0, &top, 0, 0, 0}
	, {"resent-reply-to", 0, &top, 0, 0, 0}
	, {"resent-to", 0, &top, 0, 0, 0}
	, {"resent-cc", 0, &top, 0, 0, 0}
	, {"resent-bcc", 0, &top, 0, 0, 0}
	, {"resent-date", 0, &top, 0, 0, 0}
	, {"resent-message-id", 0, &top, 0, 0, 0}
	, {"content-length", 0, 0, 0, 0, 0}
	, {0, 0, &bottom, 0, 0, 0}
};

void
envelope_make()
{
	if (recipstrategy != 'a')
	{
		if (flagenveloperecipients)
			recipient_addlist(&enveloperecipients);
		else
		{
			recipient_addlist(&to);
			recipient_addlist(&cc);
			recipient_addlist(&bcc);
		}
	}
	if (!sender.len)
		sender_get(&envelopesender);
	if (!sender.len)
		sender_get(&returnpath);
	if (!sender.len)
	{
		if (!rewritehost_addr(&sender, config_data(&suser)->s, config_data(&suser)->len, config_data(&rewrite)))
			nomem();
		if (flagverprecip && !stralloc_cats(&sender, "-@[]"))
			nomem();
		if (!stralloc_0(&sender))
			nomem();
	}
}

void
envelope_print()
{
	int             i;
	int             j;

	myputs("Envelope-Sender: ");
	if (!mess822_quote(&tmp, sender.s, (char *) 0))
		nomem();
	myput(tmp.s, tmp.len);
	myputs("\n");
	myputs("Envelope-Recipients:\n");
	for (j = i = 0; j < recipients.len; ++j)
	{
		if (!recipients.s[j])
		{
			if (!mess822_quote(&tmp, recipients.s + i, (char *) 0))
				nomem();
			myputs("  ");
			myput(tmp.s, tmp.len);
			myputs(",\n");
			i = j + 1;
		}
	}
}

void
datemsgid_print()
{
	if (!date.known)
	{
		caltime_utc(&date.ct, &start, (int *) 0, (int *) 0);
		date.known = 1;
	}
	if (!mess822_date(&tmp, &date))
		nomem();
	myputs("Date: ");
	myput(tmp.s, tmp.len);
	myputs("\n");
	if (!msgid.len)
	{
		if (!stralloc_copys(&msgid, "Message-ID: <"))
			nomem();
		if (!stralloc_catlong(&msgid, date.ct.date.year))
			nomem();
		if (!stralloc_catint0(&msgid, date.ct.date.month, 2))
			nomem();
		if (!stralloc_catint0(&msgid, date.ct.date.day, 2))
			nomem();
		if (!stralloc_catint0(&msgid, date.ct.hour, 2))
			nomem();
		if (!stralloc_catint0(&msgid, date.ct.minute, 2))
			nomem();
		if (!stralloc_catint0(&msgid, date.ct.second, 2))
			nomem();
		if (!stralloc_cat(&msgid, &idappend))
			nomem();
		if (!stralloc_cats(&msgid, ">\n"))
			nomem();
	}
	myput(msgid.s, msgid.len);
}

void
from_print()
{
	putlist("From: ", &from);
	if (!from.len)
	{
		if (!rewritehost_addr(&tmp2, config_data(&user)->s, config_data(&user)->len, config_data(&rewrite)))
			nomem();
		if (!stralloc_0(&tmp2))
			nomem();
		myputs("From: ");
		if (!mess822_quote(&tmp, tmp2.s, config(&name) ? config_data(&name)->s : (char *) 0))
			nomem();
		myput(tmp.s, tmp.len);
		myputs("\n");
	}
	putlist("Sender: ", &headersender);
}

int
inmft(list)
	stralloc       *list;
{
	int             i;
	int             j;

	for (j = i = 0; j < list->len; ++j)
	{
		if (!list->s[j])
		{
			if (list->s[i] == '+' && constmap(&mapmft, list->s + i + 1, j - i - 1))
				return 1;
			i = j + 1;
		}
	}
	return 0;
}

void
response_print()
{
	putlist("Reply-To: ", &replyto);
	putlist("Mail-Reply-To: ", &mailreplyto);

	if (!followupto.len && config(&mft) && (inmft(&to) || inmft(&cc)))
	{
		if (!stralloc_cat(&followupto, &to))
			nomem();
		if (!stralloc_cat(&followupto, &cc))
			nomem();
	}
	putlist("Mail-Followup-To: ", &followupto);
}

void
to_print()
{
	if (!to.len && !cc.len)
		myputs("Cc: recipient list not shown: ;\n");
	putlist("To: ", &to);
	putlist("Cc: ", &cc);
	putlist("Notice-Requested-Upon-Delivery-To: ", &nrudt);
}

void
finishheader()
{
	if (!mess822_end(&h))
		nomem();
	if (flagqueue && qmail_open(&qq) == -1)
		strerr_die2sys(111, FATAL, "unable to run qmail-queue: ");
	if (flagkillreturnpath)
		returnpath.len = 0;
	if (flagkillmsgid)
		msgid.len = 0;
	if (flagkillfrom)
		from.len = 0;
	rewritelist(&to);
	rewritelist(&cc);
	rewritelist(&bcc);
	rewritelist(&nrudt);
	rewritelist(&from);
	rewritelist(&headersender);
	rewritelist(&replyto);
	rewritelist(&mailreplyto);
	rewritelist(&followupto);
	rewritelist(&returnpath);
	rewritelist(&enveloperecipients);
	rewritelist(&envelopesender);
	envelope_make();
	if (!flagqueue)
		envelope_print();
	myput(top.s, top.len);
	datemsgid_print();
	from_print();
	response_print();
	to_print();
	myput(bottom.s, bottom.len);
}

void
finishmessage()
{
	char           *qqx;
	int             i;
	int             j;

	if (!flagqueue)
		substdio_flush(subfdoutsmall);
	else
	{
		qmail_from(&qq, sender.s);
		for (j = i = 0; j < recipients.len; ++j)
		{
			if (!recipients.s[j])
			{
				qmail_to(&qq, recipients.s + i);
				i = j + 1;
			}
		}
		qqx = qmail_close(&qq);
		if (*qqx)
			strerr_die2x(*qqx == 'D' ? 100 : 111, FATAL, qqx + 1);
	}
	_exit(0);
}

char           *argsender = 0;
stralloc        line = { 0 };
int             match;

int
main(argc, argv)
	int             argc;
	char          **argv;
{
	int             i, opt, flagheader = 1;
	char           *qbase;
	char          **e;

	sig_pipeignore();
	tai_now(&start);
	if (config_env(&qmailinject, "QMAILINJECT") == -1)
		nomem();
	if (config(&qmailinject))
	{
		for (i = 0; i < config_data(&qmailinject)->len; ++i)
		{
			switch (config_data(&qmailinject)->s[i])
			{
			case 'F':
			case 'f':
				flagkillfrom = 1;
				break;
			case 'I':
			case 'i':
				flagkillmsgid = 1;
				break;
			case 'M':
			case 'm':
				flagverpmess = 1;
				break;
			case 'R':
			case 'r':
				flagverprecip = 1;
				break;
			case 'S':
			case 's':
				flagkillreturnpath = 1;
				break;
			}
		}
	}
	while ((opt = getopt(argc, argv, "nNaAhHFIMRSf:")) != opteof)
	{
		switch (opt)
		{
		case 'n':
			flagqueue = 0;
			break;
		case 'N':
			flagqueue = 1;
			break;
		case 'a':
		case 'A':
		case 'h':
		case 'H':
			recipstrategy = opt;
			break;
		case 'F':
			flagkillfrom = 1;
			break;
		case 'I':
			flagkillmsgid = 1;
			break;
		case 'M':
			flagverpmess = 1;
			break;
		case 'R':
			flagverprecip = 1;
			break;
		case 'S':
			flagkillreturnpath = 1;
			break;
		case 'f':
			argsender = optarg;
			break;
		default:
			strerr_die1x(100, "new-inject: usage: new-inject [ -nNaAhHFIMRS ] [ -f sender ] [ recip ... ]\n");
		}
	} /*- while ((opt = getopt(argc, argv, "nNaAhHFIMRSf:")) != opteof) */
	argv += optind;
	if (chdir(auto_sysconfdir) == -1)
		strerr_die4sys(111, FATAL, "unable to switch to ", auto_sysconfdir, ": ");
	if (leapsecs_init() == -1)
		strerr_die2sys(111, FATAL, "unable to init leapsecs: ");
	if (config_env(&fnmft, "QMAILMFTFILE") == -1)
		nomem();
	if (config(&fnmft))
	{
		if (!stralloc_0(config_data(&fnmft)))
			nomem();
		if (config_readfile(&mft, config_data(&fnmft)->s) == -1)
			strerr_die2sys(111, FATAL, "unable to read $QMAILMFTFILE: ");
		if (config(&mft))
			if (!constmap_init(&mapmft, config_data(&mft)->s, config_data(&mft)->len, 0))
				nomem();
	}
	if (config_env(&fnrewrite, "QMAILREWRITEFILE") == -1)
		nomem();
	if (config(&fnrewrite))
	{
		if (!stralloc_0(config_data(&fnrewrite)))
			nomem();
		if (config_readfile(&rewrite, config_data(&fnrewrite)->s) == -1)
			strerr_die2sys(111, FATAL, "unable to read $QMAILREWRITEFILE: ");
	}
	if (!(qbase = env_get("QUEUE_BASE")))
	{
		if (!controldir)
		{
			if (!(controldir = env_get("CONTROLDIR")))
				controldir = auto_control;
		}
		if (chdir(controldir) == -1)
			strerr_die4sys(111, FATAL, "unable to switch to ", controldir, ": ");
		if (!access("defaultqueue", X_OK))
		{
			envdir_set("defaultqueue");
			if ((e = pathexec(0)))
				environ = e;
		}
	}
	if (config_env(&name, "QMAILNAME") == -1)
		nomem();
	if (config_env(&name, "MAILNAME") == -1)
		nomem();
	if (config_env(&name, "NAME") == -1)
		nomem();
	if (config(&name) && !stralloc_0(config_data(&name)))
		nomem();
	if (config_env(&user, "QMAILUSER") == -1)
		nomem();
	if (config_env(&user, "MAILUSER") == -1)
		nomem();
	if (config_env(&user, "USER") == -1)
		nomem();
	if (config_env(&user, "LOGNAME") == -1)
		nomem();
	if (config_default(&user, "anonymous") == -1)
		nomem();
	if (config_env(&suser, "QMAILSUSER") == -1)
		nomem();
	if (config_copy(&suser, &user) == -1)
		nomem();
	if (config_env(&host, "QMAILHOST") == -1)
		nomem();
	if (config_env(&host, "MAILHOST") == -1)
		nomem();
	if (config_default(&host, "") == -1)
		nomem();
	if (rwhconfig(&rewrite, &idappend) == -1)
		strerr_die1(111, FATAL, &rwhconfig_err);
	if (config_env(&shost, "QMAILSHOST") == -1)
		nomem();
	if (config_copy(&shost, &host) == -1)
		nomem();
	if (!stralloc_copys(&sender, ""))
		nomem();
	if (argsender)
	{
		if (!rewritehost_addr(&sender, argsender, str_len(argsender), config_data(&rewrite)))
			nomem();
		if (!stralloc_0(&sender))
			nomem();
	}
	if (recipstrategy == 'A')
		recipstrategy = (*argv ? 'a' : 'h');
	if (!stralloc_copys(&recipients, ""))
		nomem();
	if (recipstrategy != 'h')
	{
		while (*argv)
			recipient_add(*argv++);
	}
	if (flagverpmess)
	{
		unsigned char   nowpack[TAI_PACK];
		unsigned long   u;

		tai_pack((char *) nowpack, &start);
		u = nowpack[4];
		u <<= 8;
		u += nowpack[5];
		u <<= 8;
		u += nowpack[6];
		u <<= 8;
		u += nowpack[7];
		if (!stralloc_cats(config_data(&suser), "-"))
			nomem();
		if (!stralloc_catulong0(config_data(&suser), u, 0))
			nomem();
		u = getpid();
		if (!stralloc_cats(config_data(&suser), "."))
			nomem();
		if (!stralloc_catulong0(config_data(&suser), u, 0))
			nomem();
	}
	if (flagverprecip && !stralloc_cats(config_data(&suser), "-"))
		nomem();
	if (!stralloc_cats(config_data(&suser), "@"))
		nomem();
	if (!stralloc_cat(config_data(&suser), config_data(&shost)))
		nomem();
	if (!stralloc_cats(config_data(&user), "@"))
		nomem();
	if (!stralloc_cat(config_data(&user), config_data(&host)))
		nomem();
	if (!mess822_begin(&h, a))
		nomem();
	for (;;)
	{
		if (getln(subfdinsmall, &line, &match, '\n') == -1)
			strerr_die2sys(111, FATAL, "unable to read input: ");
		if (!match && !line.len)
			break;
		if (!match && !stralloc_append(&line, "\n"))
			nomem();
		if (flagheader && !mess822_ok(&line))
		{
			finishheader();
			flagheader = 0;
			if (line.len != 1)
				myput("\n", 1);
		}
		if (!flagheader)
			myput(line.s, line.len);
		else
		if (!mess822_line(&h, &line))
			nomem();
		if (!match)
			break;
	}
	if (flagheader)
		finishheader();
	finishmessage();
	return(0);
}

void
getversion_new_inject_c()
{
	static char    *x = "$Id: new-inject.c,v 1.9 2016-05-21 14:48:09+05:30 Cprogrammer Exp mbhangui $";

	x++;
}
